timeutil/schedule.go

```diff
@@ -0,0 +1,234 @@
+)
+
+var validTime = regexp.MustCompile(`^([0-9]|0[0-9]|1[0-9]|2[0-3]):([0-5][0-9]):?([0-5][0-9])?$`)
+
+type TimeOfDay struct {
+	Hour   int
+	Minute int
+	Second int
+}
+
+func (td TimeOfDay) Less(other TimeOfDay) bool {
+	if td.Hour < other.Hour {
+		return true
+	}
+	if td.Minute < other.Minute {
+		return true
+	}
+	if td.Second < other.Second {
+		return true
+	}
+	return false
+}
+
+// ParseTime parses a string that contains hour:minute and returns
+// an TimeOfDay type or an error
+func ParseTime(s string) (t TimeOfDay, err error) {
+	m := validTime.FindStringSubmatch(s)
+	if len(m) == 0 {
+		return t, fmt.Errorf("cannot parse %q", s)
+	}
+	t.Hour, err = strconv.Atoi(m[1])
+	if err != nil {
+		return t, fmt.Errorf("cannot parse %q: %s", m[1], err)
+	}
+	t.Minute, err = strconv.Atoi(m[2])
+	if err != nil {
+		return t, fmt.Errorf("cannot parse %q: %s", m[2], err)
+	}
+	if m[3] != "" {
+		t.Second, err = strconv.Atoi(m[3])
+		if err != nil {
+			return t, fmt.Errorf("cannot parse %q: %s", m[3], err)
+		}
+	}
+	return t, nil
+}
+
+// Schedule defines a start and end time and an optional weekday in which
+// events should run.
+type Schedule struct {
+	Start TimeOfDay
+	End   TimeOfDay
+
+	Weekday string
+}
+
+// Matches returns true when the given time is within the schedule
+// interval.
+func (sched *Schedule) Matches(t time.Time) bool {
+	// if the schedule is limited to a specific weekday, we need
+	// to check if we are on that day
+	if sched.Weekday != "" {
+		wd := time.Weekday(weekdayMap[sched.Weekday])
+		if t.Weekday() != wd {
+			return false
+		}
+	}
+
+	// check if we are within the schedule time window
+	if t.Hour() >= sched.Start.Hour && t.Minute() >= sched.Start.Minute {
+		if t.Hour() < sched.End.Hour {
+			return true
+		}
+		if t.Hour() == sched.End.Hour && t.Minute() <= sched.End.Minute {
+			return true
+		}
+	}
+
+	return false
+}
+
+// SameInterval returns true if the given times are within the same
+// interval. Same means that they are on the same day (if its a
+// schedule that runs on every day or the same week (if its a schedule
+// that is run only on a specific weekday).
+//
+// E.g. for a schedule of "9:00-11:00"
+//
+// t1="2017-01-01 9:10", t1="2017-01-01 9:30"
+// (same day) is the same interval
+//
+// t1="2017-01-01 9:10", t1="2017-01-02 9:30"
+// (different day) is the *not* same interval
+//
+func (sched *Schedule) SameInterval(t1, t2 time.Time) bool {
+	if !sched.Matches(t1) || !sched.Matches(t2) {
+		// FIXME: or return error here?
+		return false
+	}
+
+	if sched.Weekday != "" {
+		t1Year, t1Week := t1.ISOWeek()
+		t2Year, t2Week := t2.ISOWeek()
+		return t1Year == t2Year && t1Week == t2Week
+	}
+
+	return t1.Year() == t2.Year() && t1.Month() == t2.Month() && t1.Day() == t2.Day()
+}
+
+var weekdayMap = map[string]int{
+	"sun": 0, "sunday": 0,
+	"mon": 1, "monday": 1,
+	"tue": 2, "tuesday": 2,
+	"wed": 3, "wednesday": 3,
+	"thu": 4, "thursday": 4,
+	"fri": 5, "friday": 5,
+	"sat": 6, "saturday": 6,
```

Comment:
Can we please drop support for full weekday names? Given how rich those strings are, it's nicer to read short names which people are super used to already, and it's nice to enforce consistency.